internal/p2p/peermanager.go

```diff
@@ -815,19 +913,91 @@ func (m *PeerManager) Advertise(peerID types.NodeID, limit uint16) []NodeAddress
 		addresses = append(addresses, m.options.SelfAddress)
 	}
 
-	for _, peer := range m.store.Ranked() {
+	var numAddresses int
+	var totalScore int
+	ranked := m.store.Ranked()
+	seenAddresses := map[NodeAddress]struct{}{}
+	scores := map[types.NodeID]int{}
+
+	// get the total number of possible addresses
+	for _, peer := range ranked {
 		if peer.ID == peerID {
 			continue
 		}
+		score := int(peer.Score())
+
+		totalScore += score
+		scores[peer.ID] = score
+		for addr := range peer.AddressInfo {
+			if _, ok := m.options.PrivatePeers[addr.NodeID]; !ok {
+				numAddresses++
+			}
+		}
+	}
+
+	var attempts int
+	var addedLastIteration bool
+
+	// if there
+	if numAddresses < int(limit) {
+		limit = uint16(numAddresses)
+	}
+
+	// collect addresses until we have the number requested
+	// (limit), or we've added all known addresses, or we've tried
+	// at least 256 times and the last time we iterated over
+	// remaining addresses we added no new candidates.
+	for len(addresses) < int(limit) && (attempts < 32 || !addedLastIteration) {
+		attempts++
+		addedLastIteration = false
+
+		for idx, peer := range ranked {
+			if peer.ID == peerID {
+				continue
+			}
 
-		for nodeAddr, addressInfo := range peer.AddressInfo {
 			if len(addresses) >= int(limit) {
-				return addresses
+				break
 			}
 
-			// only add non-private NodeIDs
-			if _, ok := m.options.PrivatePeers[nodeAddr.NodeID]; !ok {
-				addresses = append(addresses, addressInfo.Address)
+			for nodeAddr, addressInfo := range peer.AddressInfo {
+				if len(addresses) >= int(limit) {
+					break
+				}
+
+				// only look at each address once, by
+				// tracking a set of addresses seen
+				if _, ok := seenAddresses[addressInfo.Address]; ok {
+					continue
+				}
+
+				// only add non-private NodeIDs
+				if _, ok := m.options.PrivatePeers[nodeAddr.NodeID]; !ok {
+					// add the peer if the total number of ranked addresses is
+					// will fit within the limit, or otherwise adding
+					// addresses based on a coin flip.
+
+					// the coinflip is based on the score, commonly, but
+					// 10% of the time we'll randomly insert a "loosing"
+					// peer.
+
+					// nolint:gosec // G404: Use of weak random number generator
+					if numAddresses <= int(limit) || rand.Intn(totalScore+1) <= scores[peer.ID]+1 || rand.Intn((idx+1)*10) <= idx+1 {
```

Comment:
This coin flip is:
```
Pick a random number between 0 and totalScore +1
If this value is less than the score of the peer, select the peer. 
```
I think we'll hit `maxAttempts` a lot doing it this way since most peers will have small scores. Either we need `maxAttempts` to grow with the `limit` or just switch to the simpler coinflip logic before.

Reply:
because we add 1 to both sides, (which we have to do to avoid passing 0 to rand.Intn) the +1 isn't notable.

The extra 10% of flip, I think helps reduce the max attempts case. Anyway, I've made it == to limit, which means we'd have to get < 1 address per iteration. (and the addedLastIteration check will let us continue past that as long as we make progress in each iteration.

Reply:
Can we do 2 * limit? I think it's quite likely we'll get less than 1 per iteration.